File: 백준/Silver/4929. 수열 걷기/수열 걷기.py
import sys
input = lambda: sys.stdin.readline().rstrip()

while True:
    n, *arr1 = map(int, input().split())
    if n == 0: break
    m, *arr2 = map(int, input().split())

    arr1.reverse()
    arr2.reverse()

    ans = 0
    c = [0, 0]

    while True:
        if not arr1 or not arr2:
            ans += max(c[0] + sum(arr1), c[1] + sum(arr2))
            break

        if arr1[-1] < arr2[-1]:
            c[0] += arr1.pop()
        elif arr2[-1] < arr1[-1]:
            c[1] += arr2.pop()
        else:
            ans += max(c) + arr1.pop()
            arr2.pop()
            c = [0, 0]

    print(ans)

File: 백준/Silver/4929. 수열 걷기/수열 걷기.cc
#include <iostream>
#include <vector>

#define endl '\n'
#define FASTIO ios::sync_with_stdio(0), cin.tie(0), cout.tie(0)

using namespace std;
using vi = vector<int>;

int n, m, c1, c2, ans;
vi arr1, arr2;

int main() {
	FASTIO;

	while (true) {
		ans = c1 = c2 = 0;
		cin >> n;
		if (n == 0) break;
		arr1 = vi(n);
		while (n--) cin >> arr1[n];
		cin >> m;
		arr2 = vi(m);
		while (m--) cin >> arr2[m];

		while (true) {
			if (arr1.empty() || arr2.empty()) {
				int s1 = 0, s2 = 0;
				for (int i : arr1) s1 += i;
				for (int i : arr2) s2 += i;
				ans += max<int>(c1 + s1, c2 + s2);
				break;
			}
			if (arr1.back() < arr2.back()) {
				c1 += arr1.back();
				arr1.pop_back();
			}
			else if (arr1.back() > arr2.back()) {
				c2 += arr2.back();
				arr2.pop_back();
			}
			else {
				ans += max<int>(c1, c2) + arr1.back();
				arr1.pop_back();
				arr2.pop_back();
				c1 = c2 = 0;
			}
		}

		cout << ans << endl;
	}


	return 0;
}
